### PR TITLE
Eco news

### PR DESCRIPTION
**Pull request by taraskovaliv**

### CODE REVIEW

File: core/src/main/java/greencity/controller/EcoNewsController.java [hunk 1]
```diff
@@ -75,23 +78,24 @@ public EcoNewsController(EcoNewsService ecoNewsService) {
...
         @ApiResponse(code = 200, message = HttpStatuses.OK),
         @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
         @ApiResponse(code = 403, message = HttpStatuses.FORBIDDEN),
     })
     @GetMapping("")
-    public ResponseEntity<List<EcoNewsDto>> findAll(
-        @ApiParam(value = "Code of the needed language.",
-            defaultValue = AppConstant.DEFAULT_LANGUAGE_CODE)
+    @ApiPageable
+    public ResponseEntity<PageableDto<EcoNewsDto>> findAll(
```
Comment (vterlyha) [inline core/src/main/java/greencity/controller/EcoNewsController.java]: Looks strange why one DTO has another DTO as Generic. Probably some problems with objects naming

File: core/src/main/java/greencity/controller/EcoNewsController.java [hunk 2]
```diff
@@ -113,4 +117,28 @@ public EcoNewsController(EcoNewsService ecoNewsService) {
...
+    @ApiOperation(value = "Get eco news by tags")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(code = 403, message = HttpStatuses.FORBIDDEN),
+    })
+    @PostMapping("/tags")
+    @ApiPageable
+    public ResponseEntity<PageableDto<EcoNewsDto>> getEcoNews(@ApiIgnore Pageable page,
+                                                              @RequestBody GetEcoNewsDto getEcoNewsDto) {
```
Comment (vterlyha) [inline core/src/main/java/greencity/controller/EcoNewsController.java]: If GetEcoNewsDto is an object with search properties better naming would be SearchCriteriaEcoNewsDto

File: core/src/main/java/greencity/exception/exceptions/TagNotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package greencity.exception.exceptions;
+
+public class TagNotFoundException extends RuntimeException {
+    /**
+     * Generated javadoc, must be replaced with real one.
+     */
```
Comment (vterlyha) [inline core/src/main/java/greencity/exception/exceptions/TagNotFoundException.java]: Wrong javadoc

File: core/src/main/java/greencity/exception/handler/CustomExceptionHandler.java [hunk 4]
```diff
@@ -263,18 +264,32 @@
         return ResponseEntity.status(HttpStatus.FORBIDDEN).body(exceptionResponse);
     }
 
+    /**
+     * Method intercept exception {@link TagNotFoundException}.
+     *
+     * @param ex      Exception witch should be intercepted.
+     * @param request contain  detail about occur exception
+     * @return ResponseEntity witch  contain http status and body  with message of exception.
```
Comment (vterlyha) [inline core/src/main/java/greencity/exception/handler/CustomExceptionHandler.java]: Please, use Ctrl + A and Crtl +Alt + L for code formatting

File: core/src/main/java/greencity/mapping/AddEcoNewsDtoRequestMapper.java [hunk 5]
```diff
@@ -41,16 +52,26 @@ public AddEcoNewsDtoRequestMapper(LanguageRepository languageRepository) {
     protected EcoNews convert(AddEcoNewsDtoRequest addEcoNewsDtoRequest) {
         EcoNews ecoNews = EcoNews.builder()
             .creationDate(ZonedDateTime.now())
-            .text(addEcoNewsDtoRequest.getText())
+            .author(userRepo.findById(addEcoNewsDtoRequest.getAuthor().getId()).orElseThrow(
```
Comment (vterlyha) [inline core/src/main/java/greencity/mapping/AddEcoNewsDtoRequestMapper.java]: Converter should not have logic related to get info from DB or some business logic. This logic should live in repos or services. Converters only responsibility to map data.

File: core/src/main/java/greencity/mapping/TagDtoMapper.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import greencity.dto.tag.TagDto;
+import greencity.entity.Tag;
+import java.util.ArrayList;
+import java.util.List;
+import org.modelmapper.AbstractConverter;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TagDtoMapper extends AbstractConverter<List<Tag>, List<TagDto>> {
```
Comment (vterlyha) [inline core/src/main/java/greencity/mapping/TagDtoMapper.java]: Converter should convert Tag to TagDto and forEach must live in some other place

File: core/src/main/java/greencity/service/impl/EcoNewsServiceImpl.java [hunk 7]
```diff
@@ -94,14 +109,54 @@ public AddEcoNewsDtoResponse save(AddEcoNewsDtoRequest addEcoNewsDtoRequest, Str
...
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @author Kovaliv Taras.
+     */
+    @Override
+    public PageableDto<EcoNewsDto> find(Pageable page, GetEcoNewsDto getEcoNewsDto) {
+        List<String> tagsStrings = parseTags(getEcoNewsDto.getTags());
```
Comment (vterlyha) [inline core/src/main/java/greencity/service/impl/EcoNewsServiceImpl.java]: stream with map method can be used here. Not need of parseTags method

File: core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java [hunk 8]
```diff
@@ -29,33 +36,48 @@
     private LanguageRepository languageRepository;
 
     private Language language = new Language(1L, "en", Collections.emptyList(), Collections.emptyList(),
-        Collections.emptyList());
+            Collections.emptyList());
+
+    private EcoNewsTranslation ecoNewsTranslation = new EcoNewsTranslation(1L, language, "title", "text", null);
 
-    private EcoNewsTranslation ecoNewsTranslation = new EcoNewsTranslation(1L, language, "title", null);
+    private User author =
```
Comment (vterlyha) [inline core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java]: Better to have some test util class, for example ModelUtils, where you can prepare needed for tests models to not add this code to tests

File: core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java [hunk 9]
```diff
@@ -29,33 +36,48 @@
...
+    private EcoNews ecoNews = new EcoNews(null, ZonedDateTime.now(), "imagePath", author,
+            Collections.singletonList(ecoNewsTranslation), Collections.emptyList());
 
     private EcoNewsTranslationDto ecoNewsTranslationDto = new EcoNewsTranslationDto(
-        new LanguageRequestDto("en"), "title");
+            new LanguageRequestDto("en"), "title", "text");
+
+    private List<TagDto> tagDtos = new TagDtoMapper().convert(Collections.singletonList(new Tag()));
 
     @Test
```
Comment (vterlyha) [inline core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java]: Use Junit5

File: core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java (hunk 8 above)
Comment (vterlyha) [reply core/src/test/java/greencity/mapping/AddEcoNewsDtoRequestMapperTest.java]: It will also helps you avoid duplication in other tests

**Comment by shevtsiv** [comment]
Consider using `--amend` flag while committing requested changes multiple times with minor 
(and meaningless in the context of one commit) changes next time. Note, this flag will append the previous commit so commit hash will be recalculated accordingly.

**Pull request merged**